### ISSUE TITLE
Computing and Copying data from CPU simultaneously

### ISSUE THREAD

**Issue by Ruberik** [m4rs-mt/ILGPU#635]
### What I'd like to do

In my application, I do GPU work in batches. Currently it looks like this:
1. Copy data to GPU.
2. Process data.
3. Goto 1 until all batches are done.

I would like it to look like this:
1. Copy batch 0
2. Parallel.Invoke(Copy data for the next batch to GPU, Process data from this batch)
3. Goto 2 until all batches are done.

I can do this just fine, as long as the "process data from this batch" step doesn't require me to synchronize. If I have to synchronize, even if it's in a different stream from the copy, it apparently waits for the copy to finish, which I don't want it to do. It isn't clear to me whether this is something I'm doing wrong, or a bug, or something impossible to avoid.

### Simplified code
Here's code where I:
1. Compute, then copy, printing out how long each one takes. Then
2. Copy and compute at the same time, without any synchronizing, printing how long things take. Joy: they take as long as the slower of the two operations! It happens nicely in parallel.

I'll show the code, then the output, then how to include the synchronizing, and then the output for that.

```c#
        static void Main(string[] args) {
            // Boring setup stuff.
            using var context = Context.Create().Cuda().EnableAlgorithms().ToContext();
            var accs = Enumerable.Range(0, context.GetCudaDevices().Count).Select(i => context.CreateCudaAccelerator(i)).ToArray();
            var blases = accs.Select(acc => new CuBlas(acc, CuBlasAPIVersion.V11)).ToArray();
            try {
                Gmm0(accs[0..1], blases[0..1]);
            } finally {
                foreach (var b in blases) b.Dispose();
                foreach (var acc in accs) acc.Dispose();
            }
        }

        private static unsafe void Gmm0(CudaAccelerator[] accelerators, CuBlas[] blases) {
            var startTime = DateTime.Now;
            void LogTime(string s) { var timeSpent = DateTime.Now.Subtract(startTime).TotalSeconds; Console.WriteLine($"{DateTime.Now}\tTime to {s}: {timeSpent:0.0000}s"); startTime = DateTime.Now; }

            // We'll multiply two square matrics by each other, four times, as our computation.
            int size = 0b11 << 12;
            var As = accelerators.Select(x => x.Allocate1D<float>(size * size)).ToArray();
            var Bs = accelerators.Select(x => x.Allocate1D<float>(size * size)).ToArray();
            var Cs = accelerators.Select(x => x.Allocate1D<float>(size * size)).ToArray();

            // We'll copy a long array as our copy operation.
            int copySize = 1 << 30;
            var Ds = accelerators.Select(x => x.Allocate1D<float>(copySize)).ToArray();
            var arr = new float[copySize];  // Declare the array.
            for (int i = 0; i < arr.Length; i++) arr[i] = i;  // Actually allocate the memory for the array.

            var copyStreams = accelerators.Select(acc => acc.CreateStream()).ToArray();
            foreach (var acc in accelerators) acc.Synchronize();

            try {
                // First, do the computation and copy in series.
                LogTime("get set up");
                Parallel.For(0, accelerators.Length, i => {
                    accelerators[i].Bind();
                    for (int k = 0; k < 4; k++) {
                        blases[i].Gemm(CuBlasOperation.NonTranspose, CuBlasOperation.NonTranspose, size, size, size, 1f, As[i].AsContiguous(), size, Bs[i].AsContiguous(), size, 1f, Cs[i].AsContiguous(), size);
                        accelerators[i].DefaultStream.Synchronize();
                    }
                });
                LogTime("multiply four times");
                var copyStartTime = DateTime.Now;
                Parallel.For(0, accelerators.Length, i => {
                    Ds[i].CopyFromCPU(arr);
                });
                LogTime("copy");
                foreach (var acc in accelerators) acc.Synchronize();
                LogTime("synchronize after copy");

                // Next, do the computation and copy at the same time.
                var invokeStartTime = DateTime.Now;
                Parallel.Invoke(
                    () => {
                        Parallel.For(0, accelerators.Length, i => {
                            accelerators[i].Bind();
                            for (int k = 0; k < 4; k++) {
                                blases[i].Gemm(CuBlasOperation.NonTranspose, CuBlasOperation.NonTranspose, size, size, size, 1f, As[i].AsContiguous(), size, Bs[i].AsContiguous(), size, 1f, Cs[i].AsContiguous(), size);
                                // accelerators[i].DefaultStream.Synchronize();
                                // Console.WriteLine($"  Finished Gemm instruction {k} on {i} after {(DateTime.Now - invokeStartTime).TotalSeconds:0.000}s.");
                            }
                        });
                    },
                    () => {
                        Parallel.For(0, accelerators.Length, i => {
                            fixed (float* ptrInternal = arr) {
                                Ds[i].AsContiguous().CopyFromCPUUnsafeAsync(copyStreams[i], ref Unsafe.AsRef<float>(ptrInternal), arr.Length);
                            }
                        });
                        Console.WriteLine($"  Finished copying from CPU after {(DateTime.Now - invokeStartTime).TotalSeconds:0.000}s.");
                    });
                foreach (var acc in accelerators) acc.Synchronize();
                LogTime("multiply and copy");
            } finally {
                foreach (var x in As) x.Dispose();
                foreach (var x in Bs) x.Dispose();
                foreach (var x in Cs) x.Dispose();
                foreach (var x in Ds) x.Dispose();
            }
        }
```

Here's the output:
```
10/5/2021 4:25:43 PM    Time to get set up: 4.9058s
10/5/2021 4:25:45 PM    Time to multiply four times: 1.7369s
10/5/2021 4:25:46 PM    Time to copy: 1.3653s
10/5/2021 4:25:46 PM    Time to synchronize after copy: 0.0000s
  Finished copying from CPU after 1.370s.
10/5/2021 4:25:48 PM    Time to multiply and copy: 1.7227s
```

Next, I uncomment these lines:
```c#
                                // accelerators[i].DefaultStream.Synchronize();
                                // Console.WriteLine($"  Finished Gemm instruction {k} on {i} after {(DateTime.Now - invokeStartTime).TotalSeconds:0.000}s.");
```
This causes there to be synchronization in the middle of the computation step, and it apparently causes the whole computation to happen after the copy, despite the fact that we start the first matrix multiplication before synchronizing, and despite being on a different AcceleratorStream. Here's the output:
```
10/5/2021 4:27:57 PM    Time to get set up: 4.8540s
10/5/2021 4:27:58 PM    Time to multiply four times: 1.7492s
10/5/2021 4:28:00 PM    Time to copy: 1.3615s
10/5/2021 4:28:00 PM    Time to synchronize after copy: 0.0000s
  Finished copying from CPU after 1.359s.
  Finished Gemm instruction 0 on 0 after 1.789s.
  Finished Gemm instruction 1 on 0 after 2.219s.
  Finished Gemm instruction 2 on 0 after 2.650s.
  Finished Gemm instruction 3 on 0 after 3.080s.
10/5/2021 4:28:03 PM    Time to multiply and copy: 3.0802s
```

### The setup
Windows 10 Pro, 64-bit
ILGPU 1.0.0-rc1
ILGPU.Algorithms 1.0.0-rc1
Two Tesla P100 GPUs. Driver Version: 462.31. CUDA Version: 11.2.

**Comment by Ruberik** [comment]
Thanks for the quick response, @MoFtZ. Everything you said matches with my understanding of what's happening. I'm hoping for help to figure out how to make the Gemm operation and the copy operation happen in parallel, like they do in the case where there's no call to Synchronize().

Part of my confusion is that I don't understand why this happens in parallel:
```
Parallel.Invoke(
    () => { foreach (k in [0,1,2,3]) { stream1.doGemm(); }},
    () => { stream2.doCopy(); }
);
```
...but this happens without any parallelism at all:
```
Parallel.Invoke(
    () => { foreach (k in [0,1,2,3]) { stream1.doGemm(); stream1.Synchronize(); }},
    () => { stream2.doCopy(); }
);
```
...and how I can change the second example to make it parallel while still keeping some kind of cpu/gpu synchronization.

I can tell there's no parallelism at all, by the way, because when the instructions are given in series, they take (computation_time + copy_time) to complete; when they're given in parallel without synchronize, they take max(computation_time + copy_time) to complete; and when they're given in parallel with synchronize, they again take (computation_time + copy_time) to complete.

**Comment by MoFtZ** [comment]
A stream represents a sequential queue. Since calling `doGemm` and `doCopy` are on separate streams, they should run in parallel.

A few things to consider..

Calling `foreach (k in [0,1,2,3]) { stream1.doGemm(); ` or `foreach (k in [0,1,2,3]) { stream1.doGemm(); stream1.Synchronize();` are effectively the same. However, calling `stream1.Synchronize` could have different behaviours. The Cuda documentation for [cuCtxCreate](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__CTX.html#group__CUDA__CTX_1g65dc0012348bc84810e2103a40d8e2cf) specifies the behaviours as:
- **CU_CTX_SCHED_SPIN**: Instruct CUDA to actively spin when waiting for results from the GPU. This can decrease latency when waiting for the GPU, but may lower the performance of CPU threads if they are performing work in parallel with the CUDA thread.
- **CU_CTX_SCHED_YIELD**: Instruct CUDA to yield its thread when waiting for results from the GPU. This can increase latency when waiting for the GPU, but can increase the performance of CPU threads performing work in parallel with the GPU.
- **CU_CTX_SCHED_BLOCKING_SYNC**: Instruct CUDA to block the CPU thread on a synchronization primitive when waiting for the GPU to finish work.

These are mapped in ILGPU to `CudaAcceleratorFlags`.

Also, according to the documentation for `Parallel.Invoke`, "No guarantees are made about the order in which the operations execute or whether they execute in parallel". It could be that the second block is not actually run in parallel. Calling `stream1.Synchronize` is causing the thread to block, which may impact on `Parallel.Invoke`'s decision to launch the second block in parallel. Try swapping the order of arguments to `Parallel.Invoke`.

**Comment by Ruberik** [comment]
**TL;DR**: My updated hypothesis is that the presence of an active copy on stream A prevents the call `blas.Gemm` from returning on stream B, even though that call only attempts to enqueue the Gemm operation. See below for details.

Great feedback. I've added a ton of extra output. I also ended up adding a second copy, because I think that does something interesting, too (and because I'm working on a different machine today, which messed up all the timing). I'll put just the most relevant part of the code at the top here, and all of it at the bottom.

```
                int multiplyCount = 2;
                int copyCount = 2;
...
                // Next, do the computation and copy at the same time.
                var invokeStartTime = DateTime.Now;
                Parallel.Invoke(
                    () => {
                        Parallel.For(0, accelerators.Length, i => {
                            Console.WriteLine($"  Started compute thread {i} after {(DateTime.Now - invokeStartTime).TotalSeconds:0.000}s.");
                            accelerators[i].Bind();
                            for (int k = 0; k < multiplyCount; k++) {
                                Console.WriteLine($"  enqueueing gemm operation {k} on {i} after {(DateTime.Now - invokeStartTime).TotalSeconds:0.000}s.");
                                blases[i].Gemm(CuBlasOperation.NonTranspose, CuBlasOperation.NonTranspose, size, size, size, 1f, As[i].AsContiguous(), size, Bs[i].AsContiguous(), size, 1f, Cs[i].AsContiguous(), size);
                                Console.WriteLine($"  enqueued gemm operation {k} on {i} after {(DateTime.Now - invokeStartTime).TotalSeconds:0.000}s.");
//                                accelerators[i].DefaultStream.Synchronize();
//                                Console.WriteLine($"  finished Gemm instruction {k} on {i} after {(DateTime.Now - invokeStartTime).TotalSeconds:0.000}s.");
                            }
                        });
                    },
                    () => {
                        Parallel.For(0, accelerators.Length, i => {
                            fixed (long* ptrInternal = arr) {
                                Console.WriteLine($"  Started copy thread {i} after {(DateTime.Now - invokeStartTime).TotalSeconds:0.000}s.");
                                for (int k = 0; k < copyCount; k++) {
                                    Console.WriteLine($"  starting copy {k} on {i} after {(DateTime.Now - invokeStartTime).TotalSeconds:0.000}s.");
                                    Ds[i].AsContiguous().CopyFromCPUUnsafeAsync(copyStreams[i], ref Unsafe.AsRef<long>(ptrInternal), arr.Length);
                                    Console.WriteLine($"  finished copy {k} {i} after {(DateTime.Now - invokeStartTime).TotalSeconds:0.000}s.");
                                }
                            }
                        });
                        //Console.WriteLine($"  Finished copying from CPU after {(DateTime.Now - invokeStartTime).TotalSeconds:0.000}s.");
                    });
                foreach (var acc in accelerators) acc.Synchronize();
                LogTime("multiply and copy");

```

Note there are two lines that start with //. I'll uncomment those later. So here's what we get:

```
10/6/2021 12:18:21 PM   Time to multiply 2 times: 1.3108s
10/6/2021 12:18:22 PM   Time to copy 2 times: 1.4036s
10/6/2021 12:18:22 PM   Time to synchronize after copy: 0.0000s
  Started copy thread 0 after 0.001s.
  starting copy 0 on 0 after 0.001s.
  Started compute thread 0 after 0.001s.
  enqueueing gemm operation 0 on 0 after 0.001s.
  finished copy 0 0 after 0.704s.
  enqueued gemm operation 0 on 0 after 0.704s.
  starting copy 1 on 0 after 0.704s.
  enqueueing gemm operation 1 on 0 after 0.705s.
  finished copy 1 0 after 1.406s.
  enqueued gemm operation 1 on 0 after 1.406s.
10/6/2021 12:18:24 PM   Time to multiply and copy: 1.9871s
```

OK. So we know that each multiply takes 0.65s, and each copy takes 0.70s, when they happen serially. But see what happens here in the parallel code:
1. Copy 0 starts.
2. I make the call to blas.Gemm, in parallel, within 1ms. That call enqueues the Gemm operation, but doesn't wait for the Gemm operation to complete. Usually, it returns almost immediately.
3. Copy 0 finishes, 0.7 seconds later.
4. The call to blas.Gemm returns, _surprisingly late_.

Let's put the commented-out "synchronize" call and output back in. Here's what we get (removing the first few lines):

```
  Started copy thread 0 after 0.001s.
  starting copy 0 on 0 after 0.001s.
  Started compute thread 0 after 0.001s.
  enqueueing gemm operation 0 on 0 after 0.001s.
  finished copy 0 0 after 0.702s.
  enqueued gemm operation 0 on 0 after 0.702s.
  starting copy 1 on 0 after 0.702s.
  finished Gemm instruction 0 on 0 after 1.293s.
  enqueueing gemm operation 1 on 0 after 1.293s.
  finished copy 1 0 after 1.413s.
  enqueued gemm operation 1 on 0 after 1.413s.
  finished Gemm instruction 1 on 0 after 1.993s.
10/6/2021 12:19:16 PM   Time to multiply and copy: 1.9929s
```

This didn't actually change the runtime, because things aren't happening properly in parallel even _without_ the call to Synchronize() in this example. Here's what happens:
1. Copy 0 starts.
2. I make the call to blas.Gemm.
3. Copy 0 finishes after 0.7 seconds.
4. The call to blas.Gemm returns.
4. Copy 1 starts.
5. Gemm 0 finishes after 0.7+0.6s (about 50ms sooner than I'd expect if it had started after copy 0 finished).
6. I make the call to blas.Gemm.
6. Copy 1 finishes after 0.7 + 0.7s.
7. The call to blas.Gemm returns.
8. Gemm 1 finishes 0.6s after the call to blas.Gemm returned (about 50ms sooner than I'd expect if it had started after copy 1 finished).

So the only things that are happening in parallel are the second copy and the first Gemm call, apparently because the presence of an active copy prevents a Gemm call from starting. But if I add this line before the start of the copy loop: `while (DateTime.Now < invokeStartTime.AddSeconds(0.01)) ;` then I get this output:
```
  Started copy thread 0 after 0.001s.
  Started compute thread 0 after 0.001s.
  enqueueing gemm operation 0 on 0 after 0.001s.
  enqueued gemm operation 0 on 0 after 0.002s.
  starting copy 0 on 0 after 0.010s.
  finished Gemm instruction 0 on 0 after 0.596s.
  enqueueing gemm operation 1 on 0 after 0.596s.
  finished copy 0 0 after 0.719s.
  enqueued gemm operation 1 on 0 after 0.719s.
  starting copy 1 on 0 after 0.719s.
  finished Gemm instruction 1 on 0 after 1.356s.
  finished copy 1 0 after 1.466s.
10/6/2021 12:38:31 PM   Time to multiply and copy: 1.4667s
```

Here the first Gemm operation starts before the first copy. An active Gemm operation doesn't prevent a copy from being started, so although there's a small delay from waiting for the first copy to complete before the second Gemm operation can start, it's essentially parallelized. Of course, "make sure you enqueue all your compute work before starting any copies on a different accelerator stream" probably isn't the best outcome here.

**Comment by MoFtZ** [comment]
The default Cuda stream usually has special behaviour, however, ILGPU has turned off that behaviour.

"Before CUDA 7, the default stream is a special stream which implicitly synchronizes with all other streams on the device."
Source: https://developer.nvidia.com/blog/gpu-pro-tip-cuda-7-streams-simplify-concurrency/

To set a stream on CuBlas, there is a `SetStream` method on the CuBlas instance.

Also, not sure if this applies to your code,  it this might be affecting your parallel performance:
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#implicit-synchronization

**Comment by MoFtZ** [comment]
@Ruberik There is potentially something going wrong with the parallel copying when using `CopyFromCPUUnsafeAsync`.

On my computer, "Time to multiply and copy" is roughly 3 seconds. However, when I change `CopyFromCPUUnsafeAsync` to `CopyFromCPU`, my total time drops to 2.4 seconds.
```CSharp
Parallel.For(0, accelerators.Length, i =>
{
    //fixed (float* ptrInternal = arr)
    //{
    //    Ds[i].AsContiguous().CopyFromCPUUnsafeAsync(copyStreams[i], ref Unsafe.AsRef<float>(ptrInternal), arr.Length);
    //}
    Ds[i].AsContiguous().CopyFromCPU(copyStreams[i], arr);
});
```

**Comment by m4rs-mt** [comment]
@Ruberik Thanks again for reporting and working on this issue 👍 I am going to take a closer look within the next days. Let's see what I can find out regarding the async copy operations in the presence of `cuBlas` calls.

**Comment by m4rs-mt** [comment]
@Ruberik I've analyzed your program provided and it keeps on failing with `illegal memory accesses` related to the `computeStream`. Have you experienced similar problems?

**Comment by m4rs-mt** [comment]
We have discussed this issue several times over the past few weeks and have not been able to reproduce the performance issues yet. That's why we are closing this issue for now. If you have further information you can share with us (e.g. additional sample code) we are happy to investigate this problem further.
